libpretixsync/src/main/java/eu/pretix/libpretixsync/sync/BlockedTicketSecretSyncAdapter.kt

```diff
@@ -0,0 +1,225 @@
+package eu.pretix.libpretixsync.sync
+
+import app.cash.sqldelight.TransactionWithoutReturn
+import app.cash.sqldelight.db.QueryResult
+import eu.pretix.libpretixsync.api.ApiException
+import eu.pretix.libpretixsync.api.PretixApi
+import eu.pretix.libpretixsync.api.ResourceNotModified
+import eu.pretix.libpretixsync.sqldelight.BlockedTicketSecret
+import eu.pretix.libpretixsync.sqldelight.ResourceSyncStatus
+import eu.pretix.libpretixsync.sqldelight.SyncDatabase
+import eu.pretix.libpretixsync.sync.SyncManager.ProgressFeedback
+import org.json.JSONException
+import org.json.JSONObject
+import java.io.UnsupportedEncodingException
+import java.net.URLEncoder
+import java.util.concurrent.ExecutionException
+
+class BlockedTicketSecretSyncAdapter(
+    db: SyncDatabase,
+    fileStorage: FileStorage,
+    eventSlug: String,
+    api: PretixApi,
+    syncCycleId: String,
+    feedback: ProgressFeedback?,
+) : BaseDownloadSyncAdapter<BlockedTicketSecret, Long>(
+    db = db,
+    api = api,
+    syncCycleId = syncCycleId,
+    eventSlug = eventSlug,
+    fileStorage = fileStorage,
+    feedback = feedback,
+) {
+
+    private var firstResponseTimestamp: String? = null
+    private var rlm: ResourceSyncStatus? = null
+
+    override fun getResourceName(): String = "blockedsecrets"
+
+    override fun getId(obj: BlockedTicketSecret): Long = obj.server_id!!
+
+    override fun getId(obj: JSONObject): Long = obj.getLong("id")
+
+    override fun getJSON(obj: BlockedTicketSecret): JSONObject = JSONObject(obj.json_data!!)
+
+    override fun queryKnownIDs(): MutableSet<Long> {
+        val res = mutableSetOf<Long>()
+        db.blockedTicketSecretQueries.selectServerIdsByEventSlug(eventSlug).execute { cursor ->
+            while (cursor.next().value) {
+                val id =
+                    cursor.getLong(0) ?: throw RuntimeException("server_id column not available")
+                res.add(id)
+            }
+
+            QueryResult.Unit
+        }
+
+        return res
+    }
+
+    override fun insert(jsonobj: JSONObject) {
+        val blocked = jsonobj.getBoolean("blocked")
+        // If not blocked and not yet in our database, we don't need to save it, as we only care
+        // about blocked entries.
+        if (!blocked) {
+            return
+        }
+
+        db.blockedTicketSecretQueries.insert(
+            blocked = blocked,
+            event_slug = eventSlug,
+            json_data = jsonobj.toString(),
+            secret = jsonobj.getString("secret"),
+            server_id = jsonobj.getLong("id"),
+            updated = jsonobj.getString("updated"),
+        )
+    }
+
+    override fun update(obj: BlockedTicketSecret, jsonobj: JSONObject) {
+        // TODO: Test new behaviour. Original version had no update case
```

Comment:
There is a specialty about blocked secrets: We don't care about a BlockedTicketSecret with blocked=0. The only reason it exists on the API layer is to allow for differential sync during updates. But there is no need to store it. So instead of updating the `blocked` field to 0, we can and should just delete the local database entry.

Reply:
Correction, the old code **does** change the blocked field to false, but it never **inserts** a new row with blocked=false, and the new code does the same, so maybe this is correct this way.

Reply:
We also have the clean-up call (`db.blockedTicketSecretQueries.deleteNotBlocked()`) in `download()`. That should clean up any rows that have been updated to `blocked=false`.
The only thing I wonder is if we can simplify any of this to make it more maintainable (e.g. no special treatment in `insert()`/`update()` and just a clean-up in `download()` maybe?). The performance impact of inserting / updating a few rows too many shouldn't be too significant (?)